norns/reference/lib/timeline.md

```diff
@@ -0,0 +1,153 @@
+  }
+```
+
+See [the crow docs](/docs/crow/timeline/#reset-keyword) for more examples.
+
+### `loop` Post-Methods
+
+**Note: as of now, only `loop` functions can have post-methods applied.**
+
+Post-methods determine how a `loop` runs, or when it will stop.
+
+By default, `loop` will repeat the time-event table endlessly. You can stop the timeline at any moment with the `:stop()` method, but you can also make the looping programmatic with the following post-methods:
+
+| Syntax                                                          | Description                                                                    |
+| --------------------------------------------------------------- | ------------------------------------------------------------------------------ |
+| my_loop = timeline.loop{duration, event, ...}:unless(predicate) | A `loop` with an `unless` *post-method* will run until the predicate is `true` |
+| my_loop = timeline.loop{duration, event, ...}:times(count)      | A `loop` with a `times` *post-method* will run until the count has been met    |
+
+See [the crow docs](/docs/crow/timeline/#loop-post-methods) for more examples.
+
+## Example:  Thirty Second Song
+
+In this example, we'll showcase the following `timeline` syntax:
+
+- `simple`: a `loop` which switches between two notes every beat
+
+- `one`: a `queue`'d `score`,  which adds note flourishes on beat-centric timestamps
+  
+  - we also use our `'reset'` keyword to loop the `score`
+
+- `two`: a `queue`'d `loop`, which adds note flourishes on divisions of the clock
+
+- `cutoffs`: a `queue`'d `loop`, which increases our engine's cutoff value every quarter-of-a-beat, until it performs this task 100 `times`
+
+- `flourishes`:  a `real` `timeline` which schedules changes to our song with seconds-centric timestamps
+
+```lua
+tl = require 'timeline'
+s = require 'sequins'
+mu = require 'musicutil'
+
+engine.name = 'PolyPerc'
+
+function init()
+  root = 48
+  cutoff = 900
+  engine.cutoff(cutoff)
+
+  -- alternate between two notes:
+  simple = tl.loop{
+    1, {send_hz, mu.note_num_to_freq(root)},
+    1, {send_hz, mu.note_num_to_freq(root+11)}
+  }
+
+  -- let's queue up some flourishes:
+  one = tl.queue():score{
+    0, {send_hz, mu.note_num_to_freq(root+7)},
+    5, {send_hz, mu.note_num_to_freq(root+19)},
+    8, 'reset'
+  }
+  two = tl.queue():loop{
+    1.5, {send_hz, mu.note_num_to_freq(root+9)},
+    0.75, {send_hz, mu.note_num_to_freq(root+14)},
+  }
+  cutoffs = tl.queue():loop{
+    0.25, function() cutoff = cutoff+20 engine.cutoff(cutoff) end
+  }:times(100)
+
+  -- and schedule them with 'real' timing:
+  flourishes = tl.real{
+    0, {print, 'song starting'},
+    4, function() one:play() cutoffs:play() end,
+    9, function() one:stop() two:play() engine.release(2) end,
+    30, function() one:stop() two:stop() simple:stop() print('song ended') end
+  }
+
+end
+
+function send_hz(hz)
+  engine.hz(hz)
```

Comment:
highly suggest instead:

function send_hz(note)
  engine.hz(mu.note_num_to_freq(note))

that way you can remove a ton of typing upwards

Reply:
oh duh, lol thank you!!

Reply:
dang that was fast